File: packages/python/src/fraud_catcher/fraud_detector.py
"""
Main fraud detector class that orchestrates all algorithms.
"""

import time
from datetime import datetime
from typing import Dict, List, Optional, Any

from .models import Transaction, FraudResult, FraudDetectorConfig, DetectionRule
from .algorithms import (
    VelocityAlgorithm, VelocityConfig,
    AmountAlgorithm, AmountConfig,
    LocationAlgorithm, LocationConfig,
    DeviceAlgorithm, DeviceConfig,
    TimeAlgorithm, TimeConfig,
    MerchantAlgorithm, MerchantConfig,
    BehavioralAlgorithm, BehavioralConfig,
    NetworkAlgorithm, NetworkConfig,
    MLAlgorithm, MLConfig
)


class FraudDetector:
    """Main fraud detection orchestrator."""
    
    def __init__(self, config: FraudDetectorConfig):
        self.config = config
        self.algorithms: Dict[str, Any] = {}
        self.rules: Dict[str, DetectionRule] = {}
        self._initialize_algorithms()
        self._initialize_rules()
    
    def _initialize_algorithms(self) -> None:
        """Initialize all fraud detection algorithms."""
        # Initialize velocity algorithm
        velocity_config = VelocityConfig(
            time_window=60,  # 1 hour
            max_transactions=10,
            max_amount=5000.0
        )
        self.algorithms['velocity'] = VelocityAlgorithm(velocity_config)
        
        # Initialize amount algorithm
        amount_config = AmountConfig(
            suspicious_threshold=1000.0,
            high_risk_threshold=5000.0,
            currency_multipliers={
                'USD': 1.0,
                'EUR': 1.1,
                'GBP': 1.3,
                'JPY': 0.007
            }
        )
        self.algorithms['amount'] = AmountAlgorithm(amount_config)
        
        # Initialize location algorithm
        location_config = LocationConfig(
            max_distance_km=1000.0,
            suspicious_distance_km=100.0,
            time_window_minutes=60,
            enable_geo_fencing=False,
            trusted_locations=[]
        )
        self.algorithms['location'] = LocationAlgorithm(location_config)
        
        # Initialize device algorithm
        device_config = DeviceConfig(
            enable_fingerprinting=True,
            suspicious_device_threshold=5,
            new_device_risk_multiplier=1.5,
            device_velocity_window=60,
            max_devices_per_user=5
        )
        self.algorithms['device'] = DeviceAlgorithm(device_config)
        
        # Initialize time algorithm
        time_config = TimeConfig(
            suspicious_hours=[0, 1, 2, 3, 4, 5, 22, 23],  # Late night/early morning
            weekend_risk_multiplier=1.2,
            holiday_risk_multiplier=1.5,
            timezone_threshold=8,  # 8 hours difference
            enable_holiday_detection=True,
            custom_holidays=[]
        )
        self.algorithms['time'] = TimeAlgorithm(time_config)
        
        # Initialize merchant algorithm
        merchant_config = MerchantConfig(
            high_risk_categories=['gambling', 'adult', 'cash_advance', 'cryptocurrency'],
            suspicious_merchants=[],
            trusted_merchants=[],
            category_risk_scores={
                'electronics': 0.3,
                'grocery': 0.1,
                'gas': 0.2,
                'restaurant': 0.2,
                'travel': 0.6,
                'gambling': 0.8,
                'adult': 0.9,
                'pharmacy': 0.4,
                'jewelry': 0.7,
                'cash_advance': 0.9
            },
            merchant_velocity_window=60,
            max_transactions_per_merchant=20,
            enable_category_analysis=True,
            enable_merchant_reputation=True
        )
        self.algorithms['merchant'] = MerchantAlgorithm(merchant_config)
        
        # Initialize behavioral algorithm
        behavioral_config = BehavioralConfig(
            enable_spending_patterns=True,
            enable_transaction_timing=True,
            enable_location_patterns=True,
            enable_device_patterns=True,
            pattern_history_days=30,
            anomaly_threshold=0.7,
            enable_machine_learning=False,
            learning_rate=0.01
        )
        self.algorithms['behavioral'] = BehavioralAlgorithm(behavioral_config)
        
        # Initialize network algorithm
        network_config = NetworkConfig(
            enable_ip_analysis=True,
            enable_proxy_detection=True,
            enable_vpn_detection=True,
            enable_tor_detection=True,
            suspicious_countries=['XX', 'ZZ'],  # Placeholder country codes
            trusted_countries=['US', 'CA', 'GB', 'DE', 'FR'],
            max_connections_per_ip=10,
            ip_velocity_window=60,
            enable_geo_ip_analysis=True,
            enable_asn_analysis=True
        )
        self.algorithms['network'] = NetworkAlgorithm(network_config)
        
        # Initialize ML algorithm
        ml_config = MLConfig(
            enable_training=True,
            model_type='ensemble',
            feature_extractors=['amount', 'velocity', 'location', 'merchant', 'device', 'time'],
            training_data_size=10000,
            retrain_interval=24,  # 24 hours
            anomaly_threshold=0.5,
            enable_feature_importance=True,
            enable_model_persistence=True
        )
        self.algorithms['ml'] = MLAlgorithm(ml_config)
    
    def _initialize_rules(self) -> None:
        """Initialize detection rules."""
        # Initialize default rules with updated weights for all algorithms
        self.rules['velocity'] = DetectionRule(
            name='velocity',
            weight=0.15,
            threshold=self.config.thresholds.get('velocity', 0.8),
            enabled='velocity' in self.config.rules,
            config={}
        )
        
        self.rules['amount'] = DetectionRule(
            name='amount',
            weight=0.15,
            threshold=self.config.thresholds.get('amount', 0.9),
            enabled='amount' in self.config.rules,
            config={}
        )
        
        self.rules['location'] = DetectionRule(
            name='location',
            weight=0.15,
            threshold=self.config.thresholds.get('location', 0.7),
            enabled='location' in self.config.rules,
            config={}
        )
        
        self.rules['device'] = DetectionRule(
            name='device',
            weight=0.15,
            threshold=self.config.thresholds.get('device', 0.8),
            enabled='device' in self.config.rules,
            config={}
        )
        
        self.rules['time'] = DetectionRule(
            name='time',
            weight=0.10,
            threshold=self.config.thresholds.get('time', 0.6),
            enabled='time' in self.config.rules,
            config={}
        )
        
        self.rules['merchant'] = DetectionRule(
            name='merchant',
            weight=0.15,
            threshold=self.config.thresholds.get('merchant', 0.7),
            enabled='merchant' in self.config.rules,
            config={}
        )
        
        self.rules['behavioral'] = DetectionRule(
            name='behavioral',
            weight=0.10,
            threshold=self.config.thresholds.get('behavioral', 0.6),
            enabled='behavioral' in self.config.rules,
            config={}
        )
        
        self.rules['network'] = DetectionRule(
            name='network',
            weight=0.10,
            threshold=self.config.thresholds.get('network', 0.8),
            enabled='network' in self.config.rules,
            config={}
        )
        
        self.rules['ml'] = DetectionRule(
            name='ml',
            weight=0.20,
            threshold=self.config.thresholds.get('ml', 0.5),
            enabled='ml' in self.config.rules,
            config={}
        )
        
        # Add custom rules if provided
        if self.config.custom_rules:
            for rule in self.config.custom_rules:
                self.rules[rule.name] = rule
    
    async def analyze(self, transaction: Transaction) -> FraudResult:
        """Analyze transaction for fraud patterns."""
        start_time = time.time()
        triggered_rules: List[str] = []
        total_weighted_score = 0.0
        total_weight = 0.0
        
        # Process each enabled rule
        for rule_name, rule in self.rules.items():
            if not rule.enabled:
                continue
            
            algorithm = self.algorithms.get(rule_name)
            if not algorithm:
                print(f"Warning: Algorithm not found for rule: {rule_name}")
                continue
            
            try:
                score = await algorithm.analyze(transaction, rule)
                
                if score >= rule.threshold:
                    triggered_rules.append(rule_name)
                
                total_weighted_score += score * rule.weight
                total_weight += rule.weight
            except Exception as error:
                print(f"Error processing rule {rule_name}: {error}")
        
        # Calculate final risk score
        risk_score = total_weighted_score / total_weight if total_weight > 0 else 0.0
        is_fraudulent = risk_score >= self.config.global_threshold
        
        # Calculate confidence based on number of triggered rules
        confidence = min(len(triggered_rules) / len(self.config.rules), 1.0) if self.config.rules else 0.0
        
        processing_time = (time.time() - start_time) * 1000  # Convert to milliseconds
        
        result = FraudResult(
            transaction_id=transaction.id,
            risk_score=min(risk_score, 1.0),
            is_fraudulent=is_fraudulent,
            confidence=confidence,
            triggered_rules=triggered_rules,
            details={
                'algorithm': 'multi-algorithm',
                'processing_time': processing_time,
                'timestamp': datetime.now()
            }
        )
        
        # Add recommendations
        result.recommendations = self._generate_recommendations(result, transaction)
        
        if self.config.enable_logging:
            print(f"Fraud analysis completed for transaction {transaction.id}:", {
                'risk_score': result.risk_score,
                'is_fraudulent': result.is_fraudulent,
                'triggered_rules': result.triggered_rules,
                'processing_time': result.details['processing_time']
            })
        
        return result
    
    def _generate_recommendations(self, result: FraudResult, transaction: Transaction) -> List[str]:
        """Generate recommendations based on analysis results."""
        recommendations = []
        
        if 'velocity' in result.triggered_rules:
            recommendations.append('Consider implementing velocity-based transaction limits')
        
        if 'amount' in result.triggered_rules:
            recommendations.append('Review transaction amount thresholds and user spending patterns')
        
        if 'location' in result.triggered_rules:
            recommendations.append('Verify transaction location and check for unusual travel patterns')
        
        if result.risk_score > 0.8:
            recommendations.append('High risk transaction - consider manual review or additional verification')
        
        if result.confidence < 0.5:
            recommendations.append('Low confidence score - consider gathering additional transaction data')
        
        return recommendations
    
    # Utility methods for external access to algorithms
    def get_velocity_stats(self, user_id: str, time_window_minutes: int = 60) -> Dict[str, Any]:
        """Get velocity statistics for user."""
        velocity_algorithm = self.algorithms.get('velocity')
        if not velocity_algorithm:
            return {'count': 0, 'total_amount': 0.0}
        
        return {
            'count': velocity_algorithm.get_transaction_count(user_id, time_window_minutes),
            'total_amount': velocity_algorithm.get_total_amount(user_id, time_window_minutes)
        }
    
    def is_suspicious_amount(self, amount: float, currency: str = 'USD') -> bool:
        """Check if amount is suspicious."""
        amount_algorithm = self.algorithms.get('amount')
        if not amount_algorithm:
            return False
        
        return amount_algorithm.is_suspicious_amount(amount, currency)
    
    def is_impossible_travel(self, from_loc: Any, to_loc: Any, time_diff_minutes: int) -> bool:
        """Check if travel between locations is impossible."""
        location_algorithm = self.algorithms.get('location')
        if not location_algorithm:
            return False
        
        return location_algorithm.is_impossible_travel(from_loc, to_loc, time_diff_minutes)
    
    # Configuration update methods
    def update_threshold(self, rule_name: str, threshold: float) -> None:
        """Update threshold for a specific rule."""
        rule = self.rules.get(rule_name)
        if rule:
            rule.threshold = threshold
    
    def update_global_threshold(self, threshold: float) -> None:
        """Update global fraud threshold."""
        self.config.global_threshold = threshold
    
    def enable_rule(self, rule_name: str) -> None:
        """Enable a specific rule."""
        rule = self.rules.get(rule_name)
        if rule:
            rule.enabled = True
    
    def disable_rule(self, rule_name: str) -> None:
        """Disable a specific rule."""
        rule = self.rules.get(rule_name)
        if rule:
            rule.enabled = False
    
    def get_config(self) -> FraudDetectorConfig:
        """Get current configuration."""
        return self.config

File: core/FraudDetector.ts
import { Transaction, FraudResult, FraudDetectorConfig, DetectionRule } from './models/Transaction';
import { VelocityAlgorithm, VelocityConfig } from './algorithms/VelocityAlgorithm';
import { AmountAlgorithm, AmountConfig } from './algorithms/AmountAlgorithm';
import { LocationAlgorithm, LocationConfig } from './algorithms/LocationAlgorithm';
import { DeviceAlgorithm, DeviceConfig } from './algorithms/DeviceAlgorithm';
import { TimeAlgorithm, TimeConfig } from './algorithms/TimeAlgorithm';
import { MerchantAlgorithm, MerchantConfig } from './algorithms/MerchantAlgorithm';
import { BehavioralAlgorithm, BehavioralConfig } from './algorithms/BehavioralAlgorithm';
import { NetworkAlgorithm, NetworkConfig } from './algorithms/NetworkAlgorithm';
import { MLAlgorithm, MLConfig } from './algorithms/MLAlgorithm';

export class FraudDetector {
  private config: FraudDetectorConfig;
  private algorithms: Map<string, any> = new Map();
  private rules: Map<string, DetectionRule> = new Map();

  constructor(config: FraudDetectorConfig) {
    this.config = config;
    this.initializeAlgorithms();
    this.initializeRules();
  }

  private initializeAlgorithms(): void {
    // Initialize velocity algorithm
    const velocityConfig: VelocityConfig = {
      timeWindow: 60, // 1 hour
      maxTransactions: 10,
      maxAmount: 5000
    };
    this.algorithms.set('velocity', new VelocityAlgorithm(velocityConfig));

    // Initialize amount algorithm
    const amountConfig: AmountConfig = {
      suspiciousThreshold: 1000,
      highRiskThreshold: 5000,
      currencyMultipliers: {
        'USD': 1,
        'EUR': 1.1,
        'GBP': 1.3,
        'JPY': 0.007
      }
    };
    this.algorithms.set('amount', new AmountAlgorithm(amountConfig));

    // Initialize location algorithm
    const locationConfig: LocationConfig = {
      maxDistanceKm: 1000,
      suspiciousDistanceKm: 100,
      timeWindowMinutes: 60,
      enableGeoFencing: false,
      trustedLocations: []
    };
    this.algorithms.set('location', new LocationAlgorithm(locationConfig));

    // Initialize device algorithm
    const deviceConfig: DeviceConfig = {
      enableFingerprinting: true,
      suspiciousDeviceThreshold: 5,
      newDeviceRiskMultiplier: 1.5,
      deviceVelocityWindow: 60,
      maxDevicesPerUser: 5
    };
    this.algorithms.set('device', new DeviceAlgorithm(deviceConfig));

    // Initialize time algorithm
    const timeConfig: TimeConfig = {
      suspiciousHours: [0, 1, 2, 3, 4, 5, 22, 23], // Late night/early morning
      weekendRiskMultiplier: 1.2,
      holidayRiskMultiplier: 1.5,
      timezoneThreshold: 8, // 8 hours difference
      enableHolidayDetection: true,
      customHolidays: []
    };
    this.algorithms.set('time', new TimeAlgorithm(timeConfig));

    // Initialize merchant algorithm
    const merchantConfig: MerchantConfig = {
      highRiskCategories: ['gambling', 'adult', 'cash_advance', 'cryptocurrency'],
      suspiciousMerchants: [],
      trustedMerchants: [],
      categoryRiskScores: {
        'electronics': 0.3,
        'grocery': 0.1,
        'gas': 0.2,
        'restaurant': 0.2,
        'travel': 0.6,
        'gambling': 0.8,
        'adult': 0.9,
        'pharmacy': 0.4,
        'jewelry': 0.7,
        'cash_advance': 0.9
      },
      merchantVelocityWindow: 60,
      maxTransactionsPerMerchant: 20,
      enableCategoryAnalysis: true,
      enableMerchantReputation: true
    };
    this.algorithms.set('merchant', new MerchantAlgorithm(merchantConfig));

    // Initialize behavioral algorithm
    const behavioralConfig: BehavioralConfig = {
      enableSpendingPatterns: true,
      enableTransactionTiming: true,
      enableLocationPatterns: true,
      enableDevicePatterns: true,
      patternHistoryDays: 30,
      anomalyThreshold: 0.7,
      enableMachineLearning: false,
      learningRate: 0.01
    };
    this.algorithms.set('behavioral', new BehavioralAlgorithm(behavioralConfig));

    // Initialize network algorithm
    const networkConfig: NetworkConfig = {
      enableIPAnalysis: true,
      enableProxyDetection: true,
      enableVPNDetection: true,
      enableTorDetection: true,
      suspiciousCountries: ['XX', 'ZZ'], // Placeholder country codes
      trustedCountries: ['US', 'CA', 'GB', 'DE', 'FR'],
      maxConnectionsPerIP: 10,
      ipVelocityWindow: 60,
      enableGeoIPAnalysis: true,
      enableASNAnalysis: true
    };
    this.algorithms.set('network', new NetworkAlgorithm(networkConfig));

    // Initialize ML algorithm
    const mlConfig: MLConfig = {
      enableTraining: true,
      modelType: 'ensemble',
      featureExtractors: ['amount', 'velocity', 'location', 'merchant', 'device', 'time'],
      trainingDataSize: 10000,
      retrainInterval: 24, // 24 hours
      anomalyThreshold: 0.5,
      enableFeatureImportance: true,
      enableModelPersistence: true
    };
    this.algorithms.set('ml', new MLAlgorithm(mlConfig));
  }

  private initializeRules(): void {
    // Initialize default rules with updated weights for all algorithms
    this.rules.set('velocity', {
      name: 'velocity',
      weight: 0.15,
      threshold: this.config.thresholds.velocity || 0.8,
      enabled: this.config.rules.includes('velocity'),
      config: {}
    });

    this.rules.set('amount', {
      name: 'amount',
      weight: 0.15,
      threshold: this.config.thresholds.amount || 0.9,
      enabled: this.config.rules.includes('amount'),
      config: {}
    });

    this.rules.set('location', {
      name: 'location',
      weight: 0.15,
      threshold: this.config.thresholds.location || 0.7,
      enabled: this.config.rules.includes('location'),
      config: {}
    });

    this.rules.set('device', {
      name: 'device',
      weight: 0.15,
      threshold: this.config.thresholds.device || 0.8,
      enabled: this.config.rules.includes('device'),
      config: {}
    });

    this.rules.set('time', {
      name: 'time',
      weight: 0.10,
      threshold: this.config.thresholds.time || 0.6,
      enabled: this.config.rules.includes('time'),
      config: {}
    });

    this.rules.set('merchant', {
      name: 'merchant',
      weight: 0.15,
      threshold: this.config.thresholds.merchant || 0.7,
      enabled: this.config.rules.includes('merchant'),
      config: {}
    });

    this.rules.set('behavioral', {
      name: 'behavioral',
      weight: 0.10,
      threshold: this.config.thresholds.behavioral || 0.6,
      enabled: this.config.rules.includes('behavioral'),
      config: {}
    });

    this.rules.set('network', {
      name: 'network',
      weight: 0.10,
      threshold: this.config.thresholds.network || 0.8,
      enabled: this.config.rules.includes('network'),
      config: {}
    });

    this.rules.set('ml', {
      name: 'ml',
      weight: 0.20,
      threshold: this.config.thresholds.ml || 0.5,
      enabled: this.config.rules.includes('ml'),
      config: {}
    });

    // Add custom rules if provided
    if (this.config.customRules) {
      this.config.customRules.forEach(rule => {
        this.rules.set(rule.name, rule);
      });
    }
  }

  async analyze(transaction: Transaction): Promise<FraudResult> {
    const startTime = Date.now();
    const triggeredRules: string[] = [];
    let totalWeightedScore = 0;
    let totalWeight = 0;

    // Process each enabled rule
    for (const [ruleName, rule] of this.rules) {
      if (!rule.enabled) continue;

      const algorithm = this.algorithms.get(ruleName);
      if (!algorithm) {
        console.warn(`Algorithm not found for rule: ${ruleName}`);
        continue;
      }

      try {
        const score = await algorithm.analyze(transaction, rule);
        
        if (score >= rule.threshold) {
          triggeredRules.push(ruleName);
        }

        totalWeightedScore += score * rule.weight;
        totalWeight += rule.weight;
      } catch (error) {
        console.error(`Error processing rule ${ruleName}:`, error);
      }
    }

    // Calculate final risk score
    const riskScore = totalWeight > 0 ? totalWeightedScore / totalWeight : 0;
    const isFraudulent = riskScore >= this.config.globalThreshold;
    
    // Calculate confidence based on number of triggered rules
    const confidence = Math.min(triggeredRules.length / this.config.rules.length, 1.0);

    const processingTime = Date.now() - startTime;

    const result: FraudResult = {
      transactionId: transaction.id,
      riskScore: Math.min(riskScore, 1.0),
      isFraudulent,
      confidence,
      triggeredRules,
      details: {
        algorithm: 'multi-algorithm',
        processingTime,
        timestamp: new Date()
      }
    };

    // Add recommendations
    result.recommendations = this.generateRecommendations(result, transaction);

    if (this.config.enableLogging) {
      console.log(`Fraud analysis completed for transaction ${transaction.id}:`, {
        riskScore: result.riskScore,
        isFraudulent: result.isFraudulent,
        triggeredRules: result.triggeredRules,
        processingTime: result.details.processingTime
      });
    }

    return result;
  }

  private generateRecommendations(result: FraudResult, transaction: Transaction): string[] {
    const recommendations: string[] = [];

    if (result.triggeredRules.includes('velocity')) {
      recommendations.push('Consider implementing velocity-based transaction limits');
    }

    if (result.triggeredRules.includes('amount')) {
      recommendations.push('Review transaction amount thresholds and user spending patterns');
    }

    if (result.triggeredRules.includes('location')) {
      recommendations.push('Verify transaction location and check for unusual travel patterns');
    }

    if (result.triggeredRules.includes('device')) {
      recommendations.push('Verify device information and check for device sharing patterns');
    }

    if (result.triggeredRules.includes('time')) {
      recommendations.push('Review transaction timing and check for unusual time patterns');
    }

    if (result.triggeredRules.includes('merchant')) {
      recommendations.push('Verify merchant information and check merchant reputation');
    }

    if (result.triggeredRules.includes('behavioral')) {
      recommendations.push('Review user behavior patterns and check for anomalies');
    }

    if (result.triggeredRules.includes('network')) {
      recommendations.push('Check IP address reputation and network security');
    }

    if (result.triggeredRules.includes('ml')) {
      recommendations.push('Machine learning model detected anomalies - consider advanced analysis');
    }

    if (result.riskScore > 0.8) {
      recommendations.push('High risk transaction - consider manual review or additional verification');
    }

    if (result.confidence < 0.5) {
      recommendations.push('Low confidence score - consider gathering additional transaction data');
    }

    // Add specific recommendations based on transaction data
    if (transaction.amount > 10000) {
      recommendations.push('Large transaction amount - consider additional verification steps');
    }

    if (transaction.location && this.isInternationalTransaction(transaction)) {
      recommendations.push('International transaction - verify user location and travel plans');
    }

    if (transaction.merchantCategory && this.isHighRiskCategory(transaction.merchantCategory)) {
      recommendations.push('High-risk merchant category - additional verification recommended');
    }

    return recommendations;
  }

  private isInternationalTransaction(transaction: Transaction): boolean {
    // Simplified international transaction detection
    // In production, you'd use proper country code validation
    return transaction.location?.country !== 'US';
  }

  private isHighRiskCategory(category: string): boolean {
    const highRiskCategories = ['gambling', 'adult', 'cash_advance', 'cryptocurrency'];
    return highRiskCategories.includes(category.toLowerCase());
  }

  // Utility methods for external access to algorithms
  getVelocityStats(userId: string, timeWindowMinutes: number = 60): { count: number; totalAmount: number } {
    const velocityAlgorithm = this.algorithms.get('velocity') as VelocityAlgorithm;
    return {
      count: velocityAlgorithm.getTransactionCount(userId, timeWindowMinutes),
      totalAmount: velocityAlgorithm.getTotalAmount(userId, timeWindowMinutes)
    };
  }

  isSuspiciousAmount(amount: number, currency: string = 'USD'): boolean {
    const amountAlgorithm = this.algorithms.get('amount') as AmountAlgorithm;
    return amountAlgorithm.isSuspiciousAmount(amount, currency);
  }

  isImpossibleTravel(from: any, to: any, timeDiffMinutes: number): boolean {
    const locationAlgorithm = this.algorithms.get('location') as LocationAlgorithm;
    return locationAlgorithm.isImpossibleTravel(from, to, timeDiffMinutes);
  }

  // Configuration update methods
  updateThreshold(ruleName: string, threshold: number): void {
    const rule = this.rules.get(ruleName);
    if (rule) {
      rule.threshold = threshold;
    }
  }

  updateGlobalThreshold(threshold: number): void {
    this.config.globalThreshold = threshold;
  }

  enableRule(ruleName: string): void {
    const rule = this.rules.get(ruleName);
    if (rule) {
      rule.enabled = true;
    }
  }

  disableRule(ruleName: string): void {
    const rule = this.rules.get(ruleName);
    if (rule) {
      rule.enabled = false;
    }
  }

  getConfig(): FraudDetectorConfig {
    return { ...this.config };
  }
}
